parser.c

```diff
@@ -338,14 +356,60 @@ static struct kanshi_profile_output *parse_profile_output(
 	}
 }
 
-static struct kanshi_profile *parse_profile(struct kanshi_parser *parser) {
-	if (!parser_expect_token(parser, KANSHI_TOKEN_LBRACKET)) {
+static struct kanshi_profile_command *parse_profile_command(
+		struct kanshi_parser *parser) {
+	// Skip the 'exec' directive.
+	if (!parser_expect_token(parser, KANSHI_TOKEN_STR)) {
 		return NULL;
 	}
 
+	if (!parser_read_line(parser)) {
+		return NULL;
+	}
+
+	if (parser->tok_str_len <= 0) {
+		fprintf(stderr, "Ignoring empty command in config file on line %d\n",
+			parser->line);
+		return NULL;
+	}
+
+	struct kanshi_profile_command *command = calloc(1, sizeof(*command));
+	command->command = strdup(parser->tok_str);
+	return command;
+}
+
+static struct kanshi_profile *parse_profile(struct kanshi_parser *parser) {
 	struct kanshi_profile *profile = calloc(1, sizeof(*profile));
 	wl_list_init(&profile->outputs);
+	wl_list_init(&profile->commands);
 
+	// First parse an optional profile name
+	parser->tok_str_len = 0;
+	if (!parser_read_str(parser)) {
+		fprintf(stderr, "expected new profile, got %s\n",
+			token_type_str(parser->tok_type));
+		return NULL;
+	}
+	profile->name = (parser->tok_str_len == 0) ? NULL : strdup(parser->tok_str);
+
+	// Then parse the opening bracket
+	if (!parser_expect_token(parser, KANSHI_TOKEN_LBRACKET)) {
+		return NULL;
+	}
+
+	// Use the bracket position to generate a default profile name
+	if (profile->name == NULL) {
+		char generated_name[100];
+		int ret = snprintf(generated_name, sizeof(generated_name),
+				"<anonymous at line %d, col %d>", parser->line, parser->col);
+		if (ret >= 0 && ret < sizeof(generated_name)) {
```

Comment:
TBH I'd just drop this check and pick a big enough `generated_name` buffer so that we are sure al `%d`s fit.

Reply:
100 is big enough already. Not combination of two integers can overflow this.